Amna ejaz/agent validate

AmnaEjaz opened the pull request:
## Summary
- Added validation in all test files, test cases, however, are failing

Commit:
Fixed yaml and added resp validator in "test_activate__experiment" (not working)

mjc1283 committed:
feat: Reject requests with invalid host (#241)

Summary:

Introduce a new configuration property server.allowedHosts. When requests are received whose host doesn't match either one of these values, or the separately configured server.host, Agent rejects the request with a 404 response. The default value of server.allowedHosts is ["localhost"].
Includes code from go-chi/hostrouter for determining request host based on X-Forwarded-Host header, Forwarded header, Host header, or host portion of URL.

Test plan:

Manually tested running Agent locally
New unit tests

Issues:

OASIS-6578

Commit:
validation complete on test_activate.py

Commit:
yaml update

fscelliott committed:
Create internal docs authoring notes.md (#258)

* Create internal docs authoring notes.md

* Update internal docs authoring notes.md

* adding code comment

fscelliott committed:
Create DONT_AUTHOR_AGENT_DOCS_HERE.md (#259)

* Create DONT_AUTHOR_AGENT_DOCS_HERE.md

* Update and rename DONT_AUTHOR_AGENT_DOCS_HERE.md to 001 - DONT_AUTHOR_AGENT_DOCS_HERE.md

* Update 001 - DONT_AUTHOR_AGENT_DOCS_HERE.md

* Update 001 - DONT_AUTHOR_AGENT_DOCS_HERE.md

Commit:
code refactored

oakbani commented on api/openapi-spec/openapi.yaml:
```diff
@@ -1,7 +1,7 @@
 info:
   title: Optimizely Agent API
-  description: Optimizely Agent is a stand-alone, open-source microservice that provides major benefits over using Optimizely SDKs in certain use cases. Its REST API offers consolidated and simplified endpoints for accessing all the functionality of Optimizely Full Stack SDKs. Use this API the control experiments (such as a feature tests). For more info, see https://docs.developers.optimizely.com ...
+  description: Optimizely Agent is a stand-alone, open-source microservice that provides major benefits over using Optimizely SDKs in certain use cases. Its REST API offers consolidated and simplified endpoints for accessing all the functionality of Optimizely Full Stack SDKs. Use this API the control experiments (such as a feature tests). For more info, see https://docs.developers.optimizely.com ...
```
try to avoid only space removal changes in unchanged code. It annoys the reviewer.

oakbani commented on api/openapi-spec/openapi.yaml:
```diff
@@ -252,17 +250,16 @@ components:
           enum:
             - feature
             - experiment
+            - null
```
remove null and test again. I think it returns an empty string in some cases

oakbani commented on api/openapi-spec/openapi.yaml:
```diff
@@ -252,17 +250,16 @@ components:
           type: object
-          additionalProperties:
-            type: string
+#          additionalProperties:
```
change this to **additionalProperties: true**

oakbani commented on tests/acceptance/helpers.py:
```diff
@@ -6,6 +6,14 @@
 import requests
 
+import yaml
+from openapi_core import create_spec
```
add openapi_core in requirements.txt

oakbani commented on tests/acceptance/helpers.py:
```diff
@@ -6,6 +6,14 @@
+from openapi_core import create_spec
+from openapi_core.validation.request.validators import RequestValidator
+from openapi_core.validation.request.datatypes import (OpenAPIRequest, RequestParameters)
+from werkzeug.datastructures import ImmutableMultiDict
```
move this to top. group all openapi imports together.

oakbani commented on tests/acceptance/helpers.py:
```diff
@@ -14,6 +22,17 @@
 
 PROJECT_ROOT = os.path.dirname(os.path.abspath(__file__))
 
+os.environ['OPENAPI_YAML'] = os.path.abspath('api/openapi-spec/openapi.yaml')
```
You need to get file path from env, not the other way. 
assume that values is in os.environ['OPENAPI_YAML_FILEPATH']
and get value from this. Otherwsie, use default value.

AmnaEjaz replied:
I did not understand what you mean, how can I assume that value exists in os.environ['OPENAPI_YAML_FILEPATH'] when i didnt set any value in it. Please elaborate this one

oakbani commented on tests/acceptance/helpers.py:
```diff
@@ -6,6 +6,14 @@
+from openapi_core.validation.request.datatypes import (OpenAPIRequest, RequestParameters)
+from werkzeug.datastructures import ImmutableMultiDict
+from openapi_core.validation.response.datatypes import OpenAPIResponse
+from openapi_core.validation.response.validators import ResponseValidator
```
nit combine with Request Valdiator import.

oakbani commented on tests/acceptance/helpers.py:
```diff
@@ -122,14 +141,26 @@ def sort_response(response_dict, *args):
+    request, request_result = create_and_validate_request(ENDPOINT_ACTIVATE, 'post', payload, params)
+
+    # raise errors if request invalid
+    request_result.raise_for_errors()
```
raise_errors part should also be moved inside those methods.

oakbani commented on tests/acceptance/helpers.py:
```diff
@@ -143,5 +174,48 @@ def override_variation(sess, override_with):
                "experimentKey": "ab_test1", "variationKey": f"{override_with}"}
-    resp = sess.post(BASE_URL + ENDPOINT_OVERRIDE, json=payload)
+
+    request, request_result = create_and_validate_request(ENDPOINT_ACTIVATE, 'post', payload)
```
Shouldn't this be ENDPOINT_OVERRIDE?

oakbani commented on tests/acceptance/helpers.py:
```diff
@@ -143,5 +174,48 @@ def override_variation(sess, override_with):
     return resp
+
+
+def create_and_validate_request(endpoint, method, payload='', params=[]):
```
add doc strings to both methods

oakbani commented on tests/acceptance/test_acceptance/test_activate.py:
```diff
@@ -568,13 +632,22 @@ def test_activate_with_config(session_obj):
+    request_result.raise_for_errors()
+
+    # resp = session_obj.post(BASE_URL + ENDPOINT_ACTIVATE, params=params, json=payload)
+    resp_activate = session_obj.post(BASE_URL + ENDPOINT_ACTIVATE, params=params, json=json.loads(payload))
```
why commented out similar above?

oakbani commented on tests/acceptance/test_acceptance/test_activate.py:
```diff
@@ -244,10 +266,20 @@ def test_activate__type(session_obj, decision_type, expected_response,
     resp = session_obj.post(BASE_URL + ENDPOINT_ACTIVATE, params=params,
                             json=json.loads(payload))
 
+    response_result = create_and_validate_response(request, resp)
```
I see common code all over the place. 

 ```
   resp = session_obj.post(BASE_URL + ENDPOINT_ACTIVATE, params=params,
                            json=json.loads(payload))
                            json=json.loads(payload))

which means that we can move all of this to a helper method. and simply call that single method in each test. 
```

Commit:
Code refactored

Commit:
yaml update

Commit:
Update openapi.yaml

AmnaEjaz commented on tests/acceptance/helpers.py:
```diff
@@ -149,17 +150,7 @@ def activate_experiment(sess):
-
-    # raise errors if response invalid
-    response_result.raise_for_errors()
+    resp = create_and_validate_request_and_response(ENDPOINT_OVERRIDE, 'post', sess, str(payload), list(params))
```
there is json deserializing issue with payload, the create_and_validate_request_and_response function has json.loads(payload) and this doesn't support the payload of activate_experiment function. Tried googling the solution, need help with this.

AmnaEjaz commented on tests/acceptance/helpers.py:
```diff
@@ -175,22 +166,22 @@ def override_variation(sess, override_with):
-
-    # raise errors if response invalid
-    response_result.raise_for_errors()
+    resp = create_and_validate_request_and_response(ENDPOINT_OVERRIDE, 'post', sess, payload=str(payload))
```
similar deserializing issue with this payload

Commit:
Update openapi.yaml

Commit:
Refactored

AmnaEjaz commented on tests/acceptance/test_acceptance/test_config.py:
```diff
@@ -136,7 +139,8 @@ def test_config(session_obj):
     """
-    resp = session_obj.get(BASE_URL + ENDPOINT_CONFIG)
+    resp = create_and_validate_request_and_response(ENDPOINT_CONFIG, 'get', session_obj)
+    # config response validations failed due to invalid schema value
```
Config tests are failing due to invalid schema issues

oakbani replied:
You don't need to put this as a comment in code. Github PR comments should suffice.

AmnaEjaz commented on tests/acceptance/test_acceptance/test_overrides.py:
```diff
@@ -36,8 +40,9 @@ def test_overrides(session_obj):
-    # Confirm deafult variation is "variation_1" (activate)
+    # Confirm default variation is "variation_1" (activate)
     activating = activate_experiment(session_obj)
+    # all overrides tests failed due to deserializing or invalidschema issues
```
All overrides tests are also failing due to deseralizing (json.loads) and invalid schema issues

oakbani commented on api/openapi-spec/openapi.yaml:
```diff
@@ -252,17 +251,15 @@ components:
-          additionalProperties:
-            type: string
+          additionalProperties: true
+#            type: string
```
nit. Remove this type:string

oakbani commented on api/openapi-spec/openapi.yaml:
```diff
@@ -347,12 +344,12 @@ components:
-      type: http
-      scheme: bearer
-      bearerFormat: JWT
+#  securitySchemes:
```
did you get a chance to take a look at this? We should ideally be asserting security part too.

oakbani commented on tests/acceptance/requirements.txt:
```diff
@@ -1,4 +1,4 @@
 pytest-clarity>=0.2.0a1
 requests>=2.22.0
-
+openapi_core
```
please specify a version.

oakbani commented on tests/acceptance/helpers.py:
```diff
@@ -23,10 +23,8 @@
-
 spec_dict = None
-with open(os.environ.get('OPENAPI_YAML'), 'r') as stream:
+with open(os.getenv('OPENAPI_YAML_PATH'), 'r') as stream:
```
also have a default value in place. So that if user doesn't set a path, the default path can be used instead. Readme specifies all commands as if you are running from the agent root directory. Similarly, set a default yaml path file here from that point of view.

oakbani commented on tests/acceptance/helpers.py:
```diff
@@ -220,11 +216,12 @@ def create_and_validate_response(request, response):
 
 
-def create_and_validate_request_and_response(endpoint, method, session, payload='', params=[]):
+def create_and_validate_request_and_response(endpoint, method, session, bypass_validation=False, payload='', params=[]):
```
rename this to bypass_req_validation

oakbani commented on tests/acceptance/test_acceptance/test_activate.py:
```diff
@@ -267,7 +266,7 @@ def test_activate_403(session_override_sdk_key):
     params = {"type": "experiment"}
 
-    request, request_result = create_and_validate_request(ENDPOINT_ACTIVATE, 'post', payload, params)
+    request, request_result = create_and_validate_request(ENDPOINT_ACTIVATE, 'post', payload=payload, params=params)
```
why aren't you using **create_and_validate_request_and_response** in this test

AmnaEjaz replied:
didn't use create_and_validate_request_and_response in any test cases with **session_override_sdk_key** arg because these have additional  **with pytest.raises(requests.exceptions.HTTPError):** check

oakbani commented on tests/acceptance/test_acceptance/test_overrides.py:
```diff
@@ -130,7 +131,7 @@ def test_overrides_403(session_override_sdk_key):
                "experimentKey": "ab_test1", "variationKey": "my_new_variation"}
 
-    request, request_result = create_and_validate_request(ENDPOINT_OVERRIDE, 'post', payload)
+    request, request_result = create_and_validate_request(ENDPOINT_OVERRIDE, 'post', payload=payload)
```
why not the **create_and_validate_request_and_response** method here?

oakbani commented on tests/acceptance/test_acceptance/test_track.py:
```diff
@@ -54,7 +54,7 @@ def test_track_403(session_override_sdk_key):
     params = {"eventKey": "myevent"}
 
-    request, request_result = create_and_validate_request(ENDPOINT_TRACK, 'post', payload, params)
+    request, request_result = create_and_validate_request(ENDPOINT_TRACK, 'post', payload=payload, params=params)
```
Same query as previous.

oakbani reviewed (changes requested):
Looks good overall. Have left some comments.

msohailhussain committed:
feat: block contenttype other than application/json (#252)

Summary:

Allow only application/json as content type for requests. Other content-type are not allowed.

Test Plan:

Added unit tests

Issues
OASIS-6582

Co-authored-by: Yasir Ali <yali@folio3.com>

aliabbasrizvi committed:
feat(auth-datafiles): Introducing support for authenticated datafiles (#246)

msohailhussain committed:
fixed some issues

oakbani committed:
all tests pass except 1

oakbani committed:
all tests passing at this point

oakbani committed:
finalized

fscelliott committed:
Rename internal docs authoring notes.md to readme.md (#260)

oakbani committed:
Merge branch 'master' into AmnaEjaz/agent-validate

Commit:
updated 403 test cases
